modules/openapi-generator/src/main/java/org/openapitools/codegen/CodegenOperation.java

```diff
@@ -31,7 +31,7 @@ public class CodegenOperation {
             isRestfulIndex, isRestfulShow, isRestfulCreate, isRestfulUpdate, isRestfulDestroy,
             isRestful, isDeprecated, isCallbackRequest, uniqueItems, hasDefaultResponse = false,
             hasErrorResponseObject; // if 4xx, 5xx responses have at least one error object defined
-    public String path, operationId, returnType, returnFormat, httpMethod, returnBaseType,
+    public String path, operationId, returnType, returnFormat, httpMethod, returnBaseType, returnInnerType,
```

Comment:
what extra information does `returnInnerType` add ?

Reply:
As an example, when the `returnType` is `List<Map<String, String>>` then `returnBaseType` will be `List` but the dart code, to be type safe, will want to call `.cast<Map<String, String>>()`. Therefore, we need to expose `Map<String, String>` to the template somehow.

Reply:
why use this instead of just items property?

Reply:
The generated code in the earlier version was simply wrong. Dart would fail to compile it under certain cases.

To see where it matters, look at the template in `dart2` named `api.mustache` on line 180. In this PR, you'll see this:

```mustache
    {{#isArray}}
      final responseBody = await _decodeBodyBytes(response);
      return (await apiClient.deserializeAsync(responseBody, '{{{returnType}}}') as List)
        .cast<{{{returnInnerType}}}>()
        .{{#uniqueItems}}toSet(){{/uniqueItems}}{{^uniqueItems}}toList(){{/uniqueItems}};
    {{/isArray}}
```

Previously, it said this:

```mustache
    {{#isArray}}
      final responseBody = await _decodeBodyBytes(response);
      return (await apiClient.deserializeAsync(responseBody, '{{{returnType}}}') as List)
        .cast<{{{returnBaseType}}}>()
        .{{#uniqueItems}}toSet(){{/uniqueItems}}{{^uniqueItems}}toList(){{/uniqueItems}};
    {{/isArray}}
```

In most cases, that is fine, but consider the following example from the openApi document from the Mattermost project. There is this api endpoint:

```json
    "/teams/{team_id}/groups_by_channels": {
      "get": {
        "tags": [
          "groups"
        ],
        "summary": "Get team groups by channels",
        "description": "Retrieve the set of groups associated with the channels in the given team grouped by channel.\n\n##### Permissions\nMust have `manage_system` permission or can access only for current user\n\n__Minimum server version__: 5.11\n",
        "operationId": "GetGroupsAssociatedToChannelsByTeam",
        "parameters": [
          {
            "name": "team_id",
            "in": "path",
            "description": "Team GUID",
            "required": true,
            "schema": {
              "type": "string"
            }
          },
          {
            "name": "page",
            "in": "query",
            "description": "The page to select.",
            "schema": {
              "type": "integer",
              "default": 0
            }
          },
          {
            "name": "per_page",
            "in": "query",
            "description": "The number of groups per page.",
            "schema": {
              "type": "integer",
              "default": 60
            }
          },
          {
            "name": "filter_allow_reference",
            "in": "query",
            "description": "Boolean which filters in the group entries with the `allow_reference` attribute set.",
            "schema": {
              "type": "boolean",
              "default": false
            }
          },
          {
            "name": "paginate",
            "in": "query",
            "description": "Boolean to determine whether the pagination should be applied or not",
            "schema": {
              "type": "boolean",
              "default": false
            }
          }
        ],
        "responses": {
          "200": {
            "description": "Group list retrieval successful",
            "content": {
              "application/json": {
                "schema": {
                  "type": "array",
                  "items": {
                    "$ref": "#/components/schemas/GroupsAssociatedToChannels"
                  }
                }
              }
            }
          },
          "400": {
            "$ref": "#/components/responses/BadRequest"
          },
          "401": {
            "$ref": "#/components/responses/Unauthorized"
          },
          "403": {
            "$ref": "#/components/responses/Forbidden"
          },
          "500": {
            "$ref": "#/components/responses/InternalServerError"
          },
          "501": {
            "$ref": "#/components/responses/NotImplemented"
          }
        }
      }
    },
```

In this case, the `returnType` will be `List<Map<String, List<GroupWithSchemeAdmin>>>` but the `returnBaseType` will be `Map` which will result in the following Dart code:

```dart
      return (await apiClient.deserializeAsync(responseBody, 'List<Map<String, List<GroupWithSchemeAdmin>>>') as List)
          .cast<Map>()
          .toList();
```

However, this code will not compile because the `.cast` operation coerces all list objects to `Map` which is the same as `Map<dynamic, dynamic>` and is LESS SPECIFIC than `Map<String, List<GroupWithSchemeAdmin>>`. The compiler can't guarantee that this cast will be valid and so the code won't compile.

By adding the `returnInnerType` variable to the `java` class and by changing the mustache template, the new code will look like this:

```dart
      return (await apiClient.deserializeAsync(responseBody, 'List<Map<String, List<GroupWithSchemeAdmin>>>') as List)
          .cast<Map<String, List<GroupWithSchemeAdmin>>>()
          .toList();
```

This code will compile.

*Why not use the items property?*

It's true that `items.dataType` contains the exact same data, but that value is not available to `api.mustache` in the context where it is needed.

This value must be exposed in the same context as `returnType`.